bfvmm/src/vmcs/test/test_vmcs_intel_x64.cpp

```diff
@@ -20,3 +21,220 @@
+    mocks.OnCall(state_in, vmcs_intel_x64_state::gs_access_rights).Do(gs_access_rights);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ldtr_access_rights).Do(ldtr_access_rights);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::tr_access_rights).Do(tr_access_rights);
+
+    // Setup 64 bit state functions
+    mocks.OnCall(state_in, vmcs_intel_x64_state::cr0).Do(cr0);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::cr3).Do(cr3);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::cr4).Do(cr4);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::dr7).Do(dr7);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::rflags).Do(rflags);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::gdt_base).Do(gdt_base);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::idt_base).Do(idt_base);
+
+    mocks.OnCall(state_in, vmcs_intel_x64_state::es_base).Do(es_base);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::cs_base).Do(cs_base);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ss_base).Do(ss_base);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ds_base).Do(ds_base);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ds_base).Do(ds_base);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::fs_base).Do(fs_base);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::gs_base).Do(gs_base);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ldtr_base).Do(ldtr_base);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::tr_base).Do(tr_base);
+
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ia32_debugctl_msr).Do(ia32_debugctl_msr);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ia32_pat_msr).Do(ia32_pat_msr);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ia32_efer_msr).Do(ia32_efer_msr);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ia32_perf_global_ctrl_msr).Do(ia32_perf_global_ctrl_msr);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ia32_sysenter_cs_msr).Do(ia32_sysenter_cs_msr);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ia32_sysenter_esp_msr).Do(ia32_sysenter_esp_msr);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ia32_sysenter_eip_msr).Do(ia32_sysenter_eip_msr);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ia32_fs_base_msr).Do(ia32_fs_base_msr);
+    mocks.OnCall(state_in, vmcs_intel_x64_state::ia32_gs_base_msr).Do(ia32_gs_base_msr);
+
+    mocks.OnCall(state_in, vmcs_intel_x64_state::dump).Do(dump);
+}
+
+static void
+setup_vmcs_intrinsics(MockRepository &mocks, memory_manager *mm, intrinsics_intel_x64 *in)
+{
+    // Emulate the memory manager
+    mocks.OnCallFunc(memory_manager::instance).Return(mm);
+    mocks.OnCallOverload(mm, (uintptr_t(memory_manager::*)(void *))&memory_manager::virt_to_phys).Do(virt_to_phys_ptr);
+
+    // Setup MSR and vmread returns to mock a successful vmcs_intel_x64::filter_unsupported
+    mocks.OnCall(in, intrinsics_intel_x64::read_msr).With(IA32_VMX_BASIC_MSR).Return(0x7fffFFFF);
+    mocks.OnCall(in, intrinsics_intel_x64::read_msr).With(IA32_VMX_TRUE_PINBASED_CTLS_MSR).Return(0x55555555aaaaAAAA);
```

Comment:
Still missing read_msr / write_msr functions for these. You can drop the `With()`, implement a read_msr / write_msr with `Do()` function and return the value you need for that specific test. The "setup" should provide a successful test case, and then each test only modifies one thing (i.e. the error your testing for).